src/dreyfus_config.erl

```diff
@@ -0,0 +1,69 @@
+-module(dreyfus_config).
```

Comment:
Did you consider using "data providers" feature from `couch_epi`?
It generates module for you and it regenerates it when it is updated.
It would also use one function clause per blacklisted element. Effectively giving you hashing structure your mentioned in the description of the PR.

Here is a quick and dirty manual test to illustrate how it works:
```
1> Handle = couch_epi_data_gen:get_handle({dreyfus, config}).
couch_epi_data_gen_dreyfus_black_list
2> couch_epi_data_gen:generate(Handle, [
    {black_list, [
        {["mydb", "myddocid", "myindexname"], true},
        {["mydb2", "myddocid2", "myindexname2"], true}
    ]}
]).
ok
3>couch_epi_data_gen:get(Handle, black_list, ["mydb", "myddocid", "myindexname"]).
true
4>couch_epi_data_gen:get(Handle, black_list, ["mydb2", "myddocid2", "myindexname2"]).
true
5> couch_epi_data_gen:get(Handle, black_list, ["mydb_miss", "myddocid", "myindexname"]).
undefined
```

In order to plug this feature properly to your code you would need to.
1. define data provider module
   ```
    -module(dreyfus_config).
    -export([data/0]).
    data() ->
         [{K, true} || K <- config:get("dreyfus", "black_list", [])].
   ```

2. Add data provider to epi config module (the one which implements `couch_epi_plugin` behaviour)
    ```
    data_providers() ->
       [
           {{dreyfus, black_list}, {module, dreyfus_config}, [{interval, 100}]}
       ].
    ```

3. If you want to get notification when blacklist changes you would need to also define a subscriber for the events and notify callback function
    ```
    data_subscriptions() ->
         [{dreyfus, black_list}].

    notify(Key, OldValue, NewValue) ->
         % do something here
         ok. 
    ```

This would eliminate most of the code in the PR.

Reply:
@iilyak : From my understanding, ```data/0``` inside the ```dreyfus_config``` module is constantly being called. I set the interval to 2.5 seconds, and my debug output shows ```config:get``` being called inside ```data/0``` every 2.5 seconds based on the code above. It's my understanding from what you said that if ```data/0``` returns a different value, then the module is regenerated. Unfortunately, I am not seeing this behavior happen happen. 

In the supervisor init add ```dreyfus_config:init```

``
init(_Args) ->
    ok = dreyfus_config:init(),
    Children = [
        child(dreyfus_index_manager)
    ],
    {ok, {{one_for_one,10,1},
        couch_epi:register_service(dreyfus_epi, Children)}}.
```
This simply does:
```
init() ->
    Handle = couch_epi_data_gen:get_handle({dreyfus, config}),
    couch_epi_data_gen:generate(Handle, [{black_list, []}]).
```
This I would assume generates an empty module ```couch_epi_data_gen_dreyfus_black_list``` with an empty list as a value.
When I call ```config:set/3```,  I see the new value appear in ```data/0```, but when
I call ```couch_epi_data_gen:get(Handle, black_list, Key).```, the value is still undefined.
I think the module is not being regenerated to contain a clause for the new Key. I even tried
on in a remsh session

```
(node1@127.0.0.1)1>  Handle = couch_epi_data_gen:get_handle({dreyfus, config}).
couch_epi_data_gen_dreyfus_config
(node1@127.0.0.1)2> couch_epi_data_gen:generate(Handle, [
(node1@127.0.0.1)2>     {black_list, [
(node1@127.0.0.1)2>         {["mydb", "myddocid", "myindexname"], true},
(node1@127.0.0.1)2>         {["mydb2", "myddocid2", "myindexname2"], true}
(node1@127.0.0.1)2>     ]}
(node1@127.0.0.1)2> ]).
ok
(node1@127.0.0.1)3> couch_epi_data_gen:get(Handle, black_list, ["mydb", "myddocid", "myindexname"]).
true
(node1@127.0.0.1)4> rpc:multicall(config, set, ["dreyfus", "black_list", [["testdb1", "6a9805a78876be4477e00de79b972a413059efae", "index-all"]]]). 
{[ok,ok,ok],[]}
(node1@127.0.0.1)5> 
(node1@127.0.0.1)5> couch_epi_data_gen:get(Handle, black_list, ["testdb1", "6a9805a78876be4477e00de79b972a413059efae", "index-all"]).
undefined
(node1@127.0.0.1)6> couch_epi_data_gen:get(Handle, black_list, ["mydb2", "myddocid2", "myindexname2"]).
true
```

I wonder if my approach is somewhat flawed.

Reply:
@iilyak : I cherry-picked your commits from https://github.com/apache/couchdb/pull/1432/files but I am still getting a similar behavior as above. The difference now is that notify is being called.